### PR TITLE
display: add UC1701 support

### PR DESCRIPTION
**Pull request by Arksine**
Adds support for the UC1701 graphics controller.  Included is font8x14.py, which contains the same VGA font in seabios with 2 bytes of padding (1 top, 1 bottom) to work with an 8x16 grid.  

Also updates config-extras.cfg and printer-creality-ender2-2017.cfg to include uc1701 support.

Signed-off-by: Eric Callahan <arksine.code@gmail.com>

### CODE REVIEW

**Comment by KevinOConnor** [comment]
Thanks!

BTW, why the different alignment on the uc1701 percent display?
I think refactoring the display code so more is shared between the different displays might have some benefits.  Not sure how best to do that though.

Separately, would the seabios 8x16 font have been an option?  (Just curious.)

File: klippy/extras/display/uc1701.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        self.vram = ([bytearray(128) for i in range(8)], 
+                    [bytearray('~'*128) for i in range(8)])
+    def build_config(self):
+        self.mcu.add_config_cmd(
+            "config_spi_without_cs oid=%d bus=%d mode=%d rate=%d shutdown_msg=" % (
+                self.spi_oid, 0, 0, 4000000))
+        self.mcu.add_config_cmd(
+            "config_digital_out oid=%d pin=%s value=%d default_value=%d max_duration=%d" % (
+                self.cs_oid, self.pins[0], 1, 1, 0))
+        self.mcu.add_config_cmd(
```
Comment (KevinOConnor) [inline klippy/extras/display/uc1701.py]: Looking a little closer at this, the use of config_spi_without_cs could be problematic.  The issue is that the commands aren't guaranteed to be sent atomically - so, for example, an spi_send to an unrelated device might get sent between the uc1701 cs enable and the uc1701 cs disable.  Then, the command to the unrelated device would be read erroneously by the uc1701.

Does the display work if the regular config_spi command is used?  Also, the spec seems to say a 10mhz speed is valid - any reason to go with the slower 4mhz speed?

**Comment by Arksine** [comment]
> BTW, why the different alignment on the uc1701 percent display?
I think refactoring the display code so more is shared between the different displays might have some benefits. Not sure how best to do that though.

Center alignment left justifies strings smaller than 4 chars when rendering direct, so it looks a bit off when the percent sign shifts.  Based on some pictures others posted it seems like the st7920 automatically right justifies them, so I was using that as my guide as to what the display should look like.

> Separately, would the seabios 8x16 font have been an option? (Just curious.)

I'm sure I could have used it.  Another situation where I was attempting to imitate what I saw on the st7920.  The extra padding does help with the percentage inside the progress bar.

File: klippy/extras/display/uc1701.py (hunk 1 above)
Comment (Arksine) [reply klippy/extras/display/uc1701.py]: Regular config_spi with spi_transfer should work.  

I couldn't find a recommended SPI speed in the datasheet, I must have missed it.   I can try 10mhz.  I'll make these changes and send them ASAP.

Comment (KevinOConnor) [reply klippy/extras/display/uc1701.py]: FYI, on page 41 of the specification I have (under AC characteristics) I read the serial bus cycle time as min 100ns (with 2.5V-3.3V power).  So, I interpreted that as 10Mhz.

Thanks.
-Kevin

**Comment by KevinOConnor** [comment]
> Center alignment left justifies strings smaller than 4 chars when rendering direct, so it looks a bit off when the percent sign shifts. Based on some pictures others posted it seems like the st7920 automatically right justifies them, so I was using that as my guide as to what the display should look like.

There's nothing special about the st7920 rendering.  It will render the exact string provided to it.

I'm guessing you're seeing that the percent sign on feed rate and fan isn't always on the right - that's also true on the st7920:

img_20180628_103016

So, I'd say that should be changed to right justification on both st7920 and uc1701.  But, center justification would be better for both on the print progress.

Or, did I miss something?

-Kevin

**Comment by Arksine** [comment]
Interesting.  Here is a picture I pulled from issue  #2  to use as a template:
screen_display_example

And here is what it looks like for the Ender 2 when center justification is used:
img_20180628_113006

Maybe whomever posted that played with the justification?

Regardless, I agree with your assessment.  Fan speed and feed rate should be right justified, while Progress should be centered.  I can submit a patch for that real quick.

**Comment by KevinOConnor** [comment]
The picture was probably from an earlier version of the code - the formatting changed around a few times.

Thanks.
-Kevin

**Pull request merged**